controllers/managedfusionoffering_controller.go

```diff
@@ -0,0 +1,82 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	v1alpha1 "github.com/red-hat-storage/managed-fusion-agent/api/v1alpha1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	controller "sigs.k8s.io/controller-runtime/pkg/controller"
+)
+
+// ManagedFusionOfferingReconciler reconciles a ManagedFusionOffering object
+type ManagedFusionOfferingReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+
+	managedFusionOffering *v1alpha1.ManagedFusionOffering
+}
+
+//+kubebuilder:rbac:groups=misf.ibm.com,resources=managedfusionofferings,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=misf.ibm.com,resources=managedfusionofferings/status,verbs=get;update;patch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// TODO(user): Modify the Reconcile function to compare the state specified by
+// the ManagedFusionOffering object against the actual cluster state, and then
+// perform operations to make the cluster state reflect the state specified by
+// the user.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.6.4/pkg/reconcile
+func (r *ManagedFusionOfferingReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("managedfusionoffering", req.NamespacedName)
+	log.Info("starting reconcile for ManangedFusionOffering")
+
+	managedFusionOfferingList := &v1alpha1.ManagedFusionOfferingList{}
+	if err := r.Client.List(ctx, managedFusionOfferingList); err != nil {
+		return ctrl.Result{}, err
+	}
+	if len(managedFusionOfferingList.Items) == 0 {
+		r.Log.Info("no managedFusionOffering CR found")
+		return ctrl.Result{}, nil
+	}
+	for i := range managedFusionOfferingList.Items {
+		r.managedFusionOffering = &managedFusionOfferingList.Items[i]
+		switch r.managedFusionOffering.Spec.Kind {
+		case v1alpha1.KindDataFoundation:
+		case v1alpha1.KindDataFoundationClient:
+		default:
+			return ctrl.Result{}, fmt.Errorf("Unknown managedFusionOffering Kind: %v", r.managedFusionOffering.Spec.Kind)
+		}
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *ManagedFusionOfferingReconciler) SetupWithManager(mgr ctrl.Manager, ctrlOptions *controller.Options) error {
```

Comment:
Can we please move the `SetupWithManager` above `reconcile`?

Reply:
Updated